### PR TITLE
Add new `Lint/RequireRangeParentheses` cop

### PR DESCRIPTION
**Pull request by koic**
## Context

It emulates the following Ruby warning.

```console
% cat example.rb
1...
2
```

```consle
% ruby example.rb
example.rb:1: warning: ... at EOL, should be parenthesized?
```

So, this cop will detect case like #10789.

## Summary

It checks that a range literal is enclosed in parentheses when the end of the range is at a line break.

### example

```ruby
# bad - Represents `(1..42)`, not endless range.
1..
42

# good - It's incompatible, but your intentions when using endless range may be:
(1..)
42

# good
1..42

# good
(1..42)

# good
(1..
42)
```

NOTE: The following is maybe intended for `(42..)`. But, compatible is `42..do_something`. So, this cop does not provide autocorrection because it is left to user.

```ruby
case condition
when 42..
  do_something
end
```

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/require_range_parentheses.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Checks that the range is enclosed in parentheses when the end of the range is line break.
+      #
+      # NOTE: The following is maybe intended for `(42..)`. But, compatible is `42..do_something`.
+      # So, this cop does not provide autocollection because it is left to user.
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/require_range_parentheses.rb:9]: autocollection -> autocorrection

File: lib/rubocop/cop/lint/require_range_parentheses.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      # ----
+      # case condition
+      # when 42..
+      #   do_something
+      # end
+      # ----
+      #
+      # @example
+      #
+      #   # bad
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/require_range_parentheses.rb:21]: You might also add comments for the bad/good cases so it's clearer what they mean.

File: lib/rubocop/cop/lint/require_range_parentheses.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Checks that the range is enclosed in parentheses when the end of the range is line break.
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/require_range_parentheses.rb:6]: that a range literal
is at a line break

File: lib/rubocop/cop/lint/require_range_parentheses.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      #
+      #   # good
+      #   1..42
+      #
+      #   # good - It's incompatible, but your intentions when using endless range may be:
+      #   (1..)
+      #   42
+      #
+      class RequireRangeParentheses < Base
+        MSG = 'Use parentheses in the endless range to avoid confusion about precedence.'
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/require_range_parentheses.rb:40]: Wrap the endless range literal %s to avoid precedence ambiguity.

**Comment by koic** [comment]
I've updated this PR. Thank you!

**Pull request merged**